packages/tss/src/tss.ts

```diff
@@ -0,0 +1,790 @@
+    const { serverEndpoints, serverPubKeys, serverThreshold, selectedServers, authSignatures } = serverOpts;
+
+    const rssClient = new RSSClient({
+      serverEndpoints,
+      serverPubKeys,
+      serverThreshold,
+      tssPubKey,
+      keyType: this._tssKeyType,
+    });
+
+    if (!this.metadata.factorPubs) throw CoreError.default(`factorPubs obj not found`);
+    if (!factorPubs) throw CoreError.default(`factorPubs not found for tssTag ${this.tssTag}`);
+    if (factorPubs.length === 0) throw CoreError.default(`factorPubs is empty`);
+
+    if (!this.metadata.tssNonces) throw CoreError.default(`tssNonces obj not found`);
+    const tssNonce: number = this.metadata.tssNonces[this.tssTag] || 0;
+
+    const oldLabel = `${verifierNameVerifierId}\u0015${this.tssTag}\u0016${tssNonce}`;
+    const newLabel = `${verifierNameVerifierId}\u0015${this.tssTag}\u0016${tssNonce + 1}`;
+
+    const { pubKey: newTSSServerPub, nodeIndexes } = await this.serviceProvider.getTSSPubKey(this.tssTag, tssNonce + 1);
+    let finalSelectedServers = selectedServers;
+
+    if (nodeIndexes?.length > 0) {
+      finalSelectedServers = nodeIndexes.slice(0, Math.min(selectedServers.length, nodeIndexes.length));
+    }
+    const refreshResponses = await rssClient.refresh({
+      factorPubs: factorPubs.map((f) => pointToHex(f)),
+      targetIndexes,
+      oldLabel,
+      newLabel,
+      sigs: authSignatures,
+      dkgNewPub: pointToHex(newTSSServerPub),
+      inputShare,
+      inputIndex,
+      selectedServers: finalSelectedServers,
+    });
+
+    const secondCommit = pointToElliptic(this._tssCurve, newTSSServerPub).add(ecPoint(this._tssCurve, tssPubKey).neg());
+    const newTSSCommits = [
+      Point.fromJSON(tssPubKey),
+      Point.fromJSON({ x: secondCommit.getX().toString(16, 64), y: secondCommit.getY().toString(16, 64) }),
+    ];
+    const factorEncs: {
+      [factorPubID: string]: FactorEnc;
+    } = {};
+    for (let i = 0; i < refreshResponses.length; i++) {
+      const refreshResponse = refreshResponses[i];
+      factorEncs[refreshResponse.factorPub.x.padStart(64, "0")] = {
+        type: "hierarchical",
+        tssIndex: refreshResponse.targetIndex,
+        userEnc: refreshResponse.userFactorEnc,
+        serverEncs: refreshResponse.serverFactorEncs,
+      };
+    }
+
+    this.metadata.addTSSData({
+      tssKeyType: this._tssKeyType,
+      tssTag: this.tssTag,
+      tssNonce: tssNonce + 1,
+      tssPolyCommits: newTSSCommits,
+      factorPubs,
+      factorEncs,
+    });
+    if (updateMetadata) await this._syncShareMetadata();
+  }
+
+  /**
+   * Derives the account nonce for the specified account index.
+   */
+  computeAccountNonce(index?: number): BN {
+    if (!index || index === 0) {
+      return new BN(0);
+    }
+
+    if (this._tssKeyType === KeyType.ed25519) {
+      throw new Error("account index not supported with ed25519");
+    }
+
+    // generation should occur during tkey.init, fails if accountSalt is absent
```

Comment:
We need to support existing user.
should we auto generate here?

Reply:
what do you mean by supporting existing user? do you think this fails if the account already exists? why?

`reconstructKey` in tkey does load the `accountSalt`, right?

Reply:
we tried it out, but didn't find a good solution because it incurred many changes in other places, so we stick with current approach for now.